File: Python/Samples/Decorator/UtDecorator.py
import unittest
from Models import Transport
from StdPricers import MilePricer, PlacePricer
from ServicePricers import Decorator, ExtraPlacePricer, HolidayPricer, DelayPricer


class UtProxy(unittest.TestCase):

    def test_scenario1(self):
        """
        標準運費：以里程計
        其他費用：加點和延遲費
        """

        transport = Transport(
            miles=200,
            place="台南",
            extraPlaceCnt=1,
            isHoliday=False,
            delayHours=3
        )

        stdPricer = MilePricer(
            customer="莉亞公主",
            receiver="反抗軍",
            freight="死星建造圖"
        )

        extraPlacePricer = ExtraPlacePricer(stdPricer)
        extraPlacePricer.price(transport)

        delayPricer = DelayPricer(extraPlacePricer)
        actual = delayPricer.price(transport)

        expected = 200 * 30 + 200 * 30 * 0.1 + 3 * 500
        self.assertEqual(actual, expected)


    def test_scenario2(self):
        """
        標準運費：以地點計
        其他費用：加點和假日運送
        """

        transport = Transport(
            miles=50,
            place="新竹",
            extraPlaceCnt=1,
            isHoliday=True,
            delayHours=0
        )

        stdPricer = PlacePricer(
            customer="莉亞公主",
            receiver="老路克天行者",
            freight="藍色光劍"
        )

        extraPlacePricer = ExtraPlacePricer(stdPricer)
        holidayPricer = HolidayPricer(extraPlacePricer)
        actual = holidayPricer.price(transport)

        expected = 1000 + (1000 * 0.1) + (1000 + 1000 * 0.1) * 0.2
        self.assertEqual(actual, expected)


if __name__ == '__main__':
    unittest.main()

File: CSharp/DP.UnitTest/UtDecorator.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using DP.Domain.Samples.Decorator;
using DP.UnitTest.Utility;
using Xunit;
using Xunit.Abstractions;

namespace DP.UnitTest
{
    public class UtDecorator
    {
        private readonly ITestOutputHelper _output;
        private readonly string _edi = string.Empty;
        private readonly string _ediExpected = string.Empty;

        public UtDecorator(ITestOutputHelper output)
        {
            this._output = output;
            System.Diagnostics.Trace.Listeners.Add(new XunitTraceListener(this._output));
        }

        [Fact]
        public void TestScenario1()
        {
            /*
             * 標準運費：以里程計
             * 其他費用：加點和延遲費
             */

            var transport = new Transport
            {
                Miles = 200,
                Place = "台南",
                ExtraPlaceCnt = 1,
                IsHoliday = false,
                DelayHours = 3
            };

            IPricer stdPricer = new MilePricer()
            {
                Customer = "莉亞公主",
                Receiver = "反抗軍",
                Freight = "死星建造圖"
            };

            IPricer extraPlacePricer = new ExtraPlacePricer(stdPricer);
            IPricer delayPricer = new DelayPricer(extraPlacePricer);
            var actual = delayPricer.Price(transport);

            var expected = 200 * 30 + 200 * 30 * 0.1 + 3 * 500;
            Assert.Equal((decimal)expected, actual);

        }

        [Fact]
        public void TestScenario2()
        {
            /*
             * 標準運費：以地點計
             * 其他費用：加點 2 和假日運送
             */

            var transport = new Transport
            {
                Miles = 50,
                Place = "新竹",
                ExtraPlaceCnt = 2,
                IsHoliday = true,
                DelayHours = 0
            };

            IPricer stdPricer = new PlacePricer()
            {
                Customer = "莉亞公主",
                Receiver = "老路克天行者",
                Freight = "藍色光劍"
            };

            IPricer extraPlacePricer = new ExtraPlacePricer(stdPricer);
            IPricer holidayPricer = new HolidayPricer(extraPlacePricer);
            var actual = holidayPricer.Price(transport);

            var expected = 1000 + (1000 * 0.1) + (1000 + 1000*0.1)*0.2;
            Assert.Equal((decimal)expected, actual);

        }

    }
}
